### PR TITLE
Fix Series.__sub__ non-nano datetime64

### LINKED ISSUE #7996: BUG: coercion of non-M8[ns] in datetime ops
```
import datetime
s = pd.Series(pd.date_range('20130101',periods=3))
s-pd.Timestamp('20130101')
s-datetime.datetime(2013,1')
```

This fails as the datetime64 is not converted properly (because numpy datetime ops suck)

```
s-np.datetime64('20130101')
```

e.g. `np.datetime64('20130101').astype('M8[ns]')` is a bug, no?

### PR DESCRIPTION
**Pull request by jbrockmendel**
The original bug report was for `Series`.  This fixes that bug and includes a test that checks for `DatetimeIndex` while we're at it.  I checked and this does _not_ fix the analogous problem in `DataFrame.  I'm hoping someone else will pick up the torch on that b/c the broadcast/dispatch is still something of a mystery to me.

- [x] closes #7996
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/ops.py [hunk 1]
```diff
@@ -671,7 +676,7 @@ def _arith_method_SERIES(op, name, str_rep, fill_zeros=None, default_axis=None,
     """
     def na_op(x, y):
         import pandas.core.computation.expressions as expressions
-
+        #
```
Comment (jreback) [inline pandas/core/ops.py]: pls don't add extraneous things

File: pandas/tests/series/test_operators.py [hunk 2]
```diff
@@ -28,6 +28,31 @@
...
+        assert dt64.dtype == 'datetime64[D]'
+        res = ser - dt64
+        expected = pd.Series([Timedelta(days=0), Timedelta(days=1),
+                              Timedelta(days=2)])
+        tm.assert_series_equal(res, expected)
+
+        res = dt64 - ser
+        tm.assert_series_equal(res, -expected)
+
+        # check for DatetimeIndex and DataFrame while we're at it
```
Comment (jreback) [inline pandas/tests/series/test_operators.py]: make the comment relevant

File: pandas/tests/series/test_operators.py [hunk 3]
```diff
@@ -28,6 +28,31 @@
...
+        dti = pd.DatetimeIndex(ser)
+        res = dti - dt64
+        tm.assert_index_equal(res, pd.Index(expected))
+
+        res = dt64 - dti
+        tm.assert_index_equal(res, pd.Index(-expected))
+
+        # TODO: This is still broken for ser.to_frame()
```
Comment (jreback) [inline pandas/tests/series/test_operators.py]: then if you want to add a test with an xfail would be helpful. (and an issue reference), TODO's are only so good.

File: pandas/tests/series/test_operators.py [hunk 4]
```diff
@@ -28,6 +28,31 @@
 from .common import TestData
 
 
+class TestDatetimeLikeArithmetic(object):
+    def test_sub_datetime64_not_ns(self):
+        # GH#7996
```
Comment (jreback) [inline pandas/tests/series/test_operators.py]: parameterize this

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
Timeout on circleCI in collection, test_geopandas in travis.  Will wait to re-push until given the go-ahead.

**Comment by jreback** [comment]
yeah don't re-push anything.

File: pandas/tests/series/test_operators.py [hunk 5]
```diff
@@ -28,6 +28,38 @@
...
+        ser = Series(date_range('20130101', periods=3))
+        dt64 = np.datetime64('2013-01-01')
```
Comment (jreback) [inline pandas/tests/series/test_operators.py]: test with a tz-aware as well (as I think that may hit your path, unless its getting caught earlier)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/series/test_operators.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/tests/series/test_operators.py]: It gets caught earlier: tz compat raises ValueError in ops._validate

File: pandas/tests/series/test_operators.py [hunk 6]
```diff
@@ -962,6 +962,36 @@ def test_timedelta64_ops_nat(self):
...
+
+        dti = pd.DatetimeIndex(ser)
+        res = dti - dt64
+        tm.assert_index_equal(res, pd.Index(expected))
+
+        res = dt64 - dti
+        tm.assert_index_equal(res, pd.Index(-expected))
+
+    @pytest.mark.xfail(reason='GH#7996 datetime64 units not converted to nano')
+    def test_frame_sub_datetime64_not_ns(self):
```
Comment (jreback) [inline pandas/tests/series/test_operators.py]: do we have a separate issue for this? if not let's create one and reference it (as closing #7996 with this PR)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/series/test_operators.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/tests/series/test_operators.py]: Just opened #18874.

File: pandas/tests/series/test_operators.py [hunk 7]
```diff
@@ -962,6 +962,36 @@ def test_timedelta64_ops_nat(self):
...
+        ser = Series(date_range('20130101', periods=3))
+        dt64 = np.datetime64('2013-01-01')
+        assert dt64.dtype == 'datetime64[D]'
+        res = ser - dt64
+        expected = pd.Series([Timedelta(days=0), Timedelta(days=1),
+                              Timedelta(days=2)])
+        tm.assert_series_equal(res, expected)
+
+        res = dt64 - ser
+        tm.assert_series_equal(res, -expected)
```
Comment (jreback) [inline pandas/tests/series/test_operators.py]: can you parametrize

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/series/test_operators.py (hunk 7 above)
Comment (jbrockmendel) [reply pandas/tests/series/test_operators.py]: Not without introducing a whole lot of extra boilerplate.

Comment (jreback) [reply pandas/tests/series/test_operators.py]: not sure that is the case. you are repeating Series/DTI testing.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/series/test_operators.py (hunk 7 above)
Comment (jbrockmendel) [reply pandas/tests/series/test_operators.py]: OK, just changed.  It is much less obvious to a casual reader exactly what this is testing.

File: pandas/tests/series/test_operators.py [hunk 8]
```diff
@@ -962,6 +962,34 @@ def test_timedelta64_ops_nat(self):
 
 
 class TestDatetimeSeriesArithmetic(object):
+    @pytest.mark.parametrize('box_cls, assert_func', [(Series,
```
Comment (jreback) [inline pandas/tests/series/test_operators.py]: this formatting is a bit atypical, put a more like

```
@pytest.mark.parametrize(
    'box, assert_func', 
    [(Series, tm.assert_series_equal),
      (pd.Index, tm.assert_index_equal)])
```
is much better, also rename ``box_cls`` -> ``box`` (consistent with elsewhere)

**Review by jreback — CHANGES_REQUESTED** [review]
small change. ping on green.

**Comment by jbrockmendel** [comment]
Ping

**Comment by jbrockmendel** [comment]
This somehow doesn't need a whatsnew merge

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**